File: source/go/atomic_counter.go
package main

import (
	"fmt"
	"sync"
	"sync/atomic"
)

func incrementWorker(counter *int32, iterations int, doneCh chan<- int, wg *sync.WaitGroup) {
	defer wg.Done()
	for i := 0; i < iterations; i++ {
		atomic.AddInt32(counter, 1)
	}
	doneCh <- 1
}

func main() {
	var counter int32
	doneCh := make(chan int, 10)
	numTasks := 10
	iterations := 100

	var wg sync.WaitGroup
	for i := 0; i < numTasks; i++ {
		wg.Add(1)
		go incrementWorker(&counter, iterations, doneCh, &wg)
	}

	wg.Wait()

	for i := 0; i < numTasks; i++ {
		<-doneCh
	}
	close(doneCh)

	finalVal := atomic.LoadInt32(&counter)
	fmt.Printf("final: %d\n", finalVal)
}

File: source/javascript/atomic_counter.js
// Lock-Free Atomic Counter
// 10 tasks each increment a shared atomic counter 100 times.
// Expected final value: 1000
async function incrementWorker(counter, iterations) {
    for (let i = 0; i < iterations; i++) {
        Atomics.add(counter, 0, 1);
    }
}

async function main() {
    const numTasks = 10;
    const iterations = 100;
    const sharedBuf = new SharedArrayBuffer(4);
    const counter = new Int32Array(sharedBuf);

    const tasks = [];
    for (let i = 0; i < numTasks; i++) tasks.push(incrementWorker(counter, iterations));
    await Promise.all(tasks);

    const finalVal = Atomics.load(counter, 0);
    console.log(`final: ${finalVal}`);
}

main();
